Add hierarchical statistics support

apatole opened the pull request:
This change adds hierarchical federated statistics support.

The existing `StatisticsController` outputs global statistics in a flat hierarchy. There are usecases where global statistics are required to be generated as per the given hierarchy configuration where NVFlare clients can be specified to belong to a particular hierarchy.

The new class `HierarchicalStatisticsController` is added to support hierarchical statistics. It is derived from `StatisticsController` and takes additional argument `hierarchy_config` for hierarchy specification file providing details about all the clients and their hierarchy.

Example hierarchy config file contents:

    {
        "Manufacturers": [
            {
                "Name": "Manufacturer-1",
                "Orgs": [
                     {
                         "Name": "Org-1",
                         "Sites": ["Site-1"]
                     },
                     {
                         "Name": "Org-2",
                         "Sites": ["Site-2"]
                     }
                ],
            },
            {
                "Name": "Manufacturer-2",
                "Orgs": [
                    {
                        "Name": "Org-3",
                        "Sites": ["Site-3", "Site-4"]
                    }
                ]
            },
        ]
    }

The above hierarchy config specifies three level hierarchy for four NVFlare clients named 'Site-1', 'Site-2', 'Site-3' and 'Site-4'. And the generate global statistics output will in hierarchical format like below. At each hierarchical level, global statistics are caulculated whereas local statistics are reported at the last hierarchical level.

```
{
    "Global": {
        <Global stats>
    },
    "Manufacturers": [
        {
            "Name": "Manufacturer-1",
            "Global": {
                <Manufacturer level stats>
            },
            "Orgs": [
                {
                    "Name": "Org-1",
                    "Global": {
                        <Org level stats>
                    },
                    "Sites": [
                        {
                            "Name": "Site-1",
                            "Local": {
                                <Local stats>
                            },
                        },
                        {
                            "Name": "Site-2",
                            "Local": {
                                <Local stats>
                            },
                        }
                    ],
                },
            ]
        },
    ],
}

```
The number of hierarchical levels and the hierarchical level names are automatically calculated from the given hierarchy config. Any number of hierarchical levels are supported.

This change also adds unit test to test hierarchical global statistics up to four hierarchical levels.

### Types of changes
<!--- Put an `x` in all the boxes that apply, and remove the not applicable items -->
- [x] Non-breaking change (fix or new feature that would not break existing functionality).
- [x] New tests added to cover the changes.
- [x] Quick tests passed locally by running `./runtest.sh`.
- [x] In-line docstrings updated.

apatole commented on nvflare/app_common/statistics/hierarchical_numeric_stats.py:
```diff
@@ -0,0 +1,477 @@
+    for metric in ordered_metrics:
+        stats[metric] = {}
+        for stat in client_metrics:
+            for device in client_metrics[stat]:
```
It's actually just the variable name and all the output is based on the hierarchical configuration. But let me just recheck this and change the variable name as well to avoid the confusion, thanks.

apatole replied:
Done, changed variable name, thanks.

yanchengnv commented on nvflare/app_common/statistics/hierarchical_numeric_stats.py:
```diff
@@ -0,0 +1,477 @@
+T = TypeVar("T")
+
+
+def get_initial_structure(client_metrics: dict, metric_task: str, ordered_metrics: dict) -> dict:
```
Please add docstrings in google style. Please give description to each arg.

apatole replied:
Done, added docstrings at all the places with `Args` and also `Returns`.

yanchengnv commented on nvflare/app_common/workflows/hierarchical_statistics_controller.py:
```diff
@@ -0,0 +1,288 @@
+            engine = fl_ctx.get_engine()
+            ws = engine.get_workspace()
+            app_dir = ws.get_app_dir(fl_ctx.get_job_id())
+            hierarchy_config_file_path = os.path.join(app_dir, "config", self.hierarchy_config)
```
Please change these two lines to:
app_conf_dir = ws.get_app_config_dir(fl_ctx.get_job_id())
hierarchy_config_file_path = os.path.join(app_conf_dir, self.hierarchy_config)

This way, you can avoid hard-coding "config". Note that the config dir's name may not be always "config".

apatole replied:
Done, thanks for the inputs, I was not aware of the `get_app_config_dir()` function.

yanchengnv commented on nvflare/app_common/workflows/hierarchical_statistics_controller.py:
```diff
@@ -0,0 +1,288 @@
+            min_clients,
+            enable_pre_run_task,
+        )
+        self.hierarchy_config = hierarchy_config
```
Please validate hierarchy_config to make sure it's a valid file name. Raise exception if not.

apatole replied:
Done, thanks.

yanchengnv commented on nvflare/app_common/workflows/hierarchical_statistics_controller.py:
```diff
@@ -0,0 +1,288 @@
+            ws = engine.get_workspace()
+            app_dir = ws.get_app_dir(fl_ctx.get_job_id())
+            hierarchy_config_file_path = os.path.join(app_dir, "config", self.hierarchy_config)
+            with open(hierarchy_config_file_path) as hierarchy_config_file:
```
The file may be missing or does not contain valid json. Please catch exception and abort (self.system_panic).

apatole replied:
Done, good catch I have added required exception handling, thanks.

yanchengnv commented on nvflare/app_common/workflows/hierarchical_statistics_controller.py:
```diff
@@ -0,0 +1,288 @@
+            abort_signal=abort_signal,
+        )
+
+        if self.hierarchy_config:
```
Since there could be errors in config, we should try to catch errors as soon as possible.
Please consider moving this block of code to start_controller(), which happens before the workflow is started.

apatole replied:
Done, I have added required exception handling and also moved this block to the start of the function. thanks.

yanchengnv reviewed (changes requested):
Nice work! I added some comments and requested changes to be more consistent with FLARE code style.

apatole commented:
Thanks @chesterxgchen and @yanchengnv for the review, I have incorporated all the review comments. Also removed few redundant source lines in the latest change, thanks

yanchengnv commented on nvflare/app_common/workflows/hierarchical_statistics_controller.py:
```diff
@@ -0,0 +1,335 @@
+        )
+
+        self.global_statistics = get_global_stats(
+            self.global_statistics, self.client_statistics, statistic_task, hierarchy_config_json
```
If self.hierarchy_config is empty/None, then hierarchy_config_json would be undefined.
It should be initialized to a default value before "if self.hierarchy_config:" or placed in the "else" clause.

apatole replied:
Done, I have added else that reports error if `self.hierarchy_config` is empty/None and returns. Hoping it looks ok now.

yanchengnv replied:
Looks good.

yanchengnv commented on nvflare/app_common/workflows/hierarchical_statistics_controller.py:
```diff
@@ -0,0 +1,335 @@
+                if isinstance(value, list):
+                    for item in value:
+                        self._recursively_round_global_stats(item)
+        if isinstance(global_stats, list):
```
Change to "elif".

apatole replied:
Sounds good, done, thanks for pointing it out.

yanchengnv commented on nvflare/app_common/workflows/hierarchical_statistics_controller.py:
```diff
@@ -0,0 +1,335 @@
+                if StC.STATS_MAX in self.global_statistics[StC.GLOBAL]:
+                    inputs[StC.STATS_MAX] = self.global_statistics[StC.GLOBAL][StC.STATS_MAX]
+
+            if tm.name == StC.STATS_VAR:
```
Consider changing to "elif".

apatole replied:
Sounds good, done, thanks for pointing it out.

yanchengnv reviewed (changes requested):
Looks better. There seems to be a logical gap about hierarchy_config_json handling (see my comments). Also added some minor style suggestions.

apatole commented:
Thanks @yanchengnv again for the review. I have incorporated review comments, hoping I addressed comments properly, let me know otherwise, thanks.

yanchengnv reviewed (dismissed):
Looks good!

apatole commented:
/build

apatole commented:
Hi @yanchengnv @chesterxgchen, Looks like someone needs to also approve workflow runs to be able to pass workflow checks and merge PR once all checks passed.

yanchengnv commented:
> > Hi @yanchengnv @chesterxgchen, Looks like someone needs to also approve workflow runs to be able to pass workflow checks and merge PR once all checks passed.
> 
> I am at conference, will take another look after that,

I already did it.

There are some format issues. 
@apatole, please run "python -m isort" and "python -m black" against your code to fix these issues.

apatole commented:
@yanchengnv no worries, thanks. Fixed one typo in copyright section in latest patch, it was introduced in earlier patch, thanks.

apatole commented:
Looks like `blossom-ci` is still waiting to be triggered.

apatole commented:
Hi @yanchengnv, please help getting the workflows triggered once and in merging the PR. Had to rebase the PR and it seems `blossom-ci` also is not getting triggered. Thanks.
